contracts/contracts/core/rlp-manager/src/main.sw

```diff
@@ -0,0 +1,663 @@
+                                              |___/
+*/
+
+mod events;
+mod errors;
+mod constants;
+
+use std::{
+    call_frames::msg_asset_id,
+    context::*,
+    revert::require,
+    primitive_conversions::u64::*,
+    math::*,
+};
+use std::hash::*;
+use core_interfaces::{
+    rlp_manager::RLPManager,
+    shorts_tracker::ShortsTracker,
+    vault::Vault,
+};
+use asset_interfaces::{
+    rlp::RLP,
+    yield_asset::YieldAsset,
+    rusd::RUSD
+};
+use src3::SRC3;
+use helpers::{
+    time::get_unix_timestamp,
+    utils::*,
+    transfer::transfer_assets,
+    zero::*,
+    reentrancy::*,
+};
+use events::*;
+use errors::*;
+use constants::*;
+
+const REVISION: u8 = 1u8;
+const DEFAULT_SUB_ID: SubId = SubId::zero();
+
+storage { 
+    // gov is not restricted to an `Address` (EOA) or a `Contract` (external)
+    // because this can be either a regular EOA (Address) or a Multisig (Contract)
+    gov: Identity = ZERO_IDENTITY,
+    lock: bool = false,
+
+    is_initialized: bool = false,
+    in_private_mode: bool = false,
+    
+    rlp_contract: ContractId = ZERO_CONTRACT,
+    vault: ContractId = ZERO_CONTRACT,
+    shorts_tracker: ContractId = ZERO_CONTRACT,
+    rusd_contract: ContractId = ZERO_CONTRACT,
+
+    cooldown_duration: u64 = 0,
+    last_added_at: StorageMap<Identity, u64> = StorageMap {},
+
+    aum_addition: u256 = 0,
+    aum_deduction: u256 = 0,
+
+    shorts_tracker_avg_price_weight: u256 = 0,
+
+    is_handler: StorageMap<Identity, bool> = StorageMap {},
+}
+
+impl RLPManager for Contract {
+    fn get_revision() -> u8 {
+        REVISION
+    }
+
+    #[storage(read, write)]
+    fn initialize(
+        vault: ContractId,
+        rusd_contract: ContractId,
+        rlp_contract: ContractId,
+        shorts_tracker: ContractId,
+        cooldown_duration: u64
+    ) {
+        require(!storage.is_initialized.read(), Error::RLPManagerAlreadyInitialized);
+        storage.is_initialized.write(true);
+
+        let gov = get_sender();
+        storage.gov.write(gov);
+        storage.rlp_contract.write(rlp_contract);
+        storage.shorts_tracker.write(shorts_tracker);
+        storage.vault.write(vault);
+        storage.rusd_contract.write(rusd_contract);
+        storage.cooldown_duration.write(cooldown_duration);
+
+        log(SetGov { gov });
+        log(SetRlpContract { rlp_contract });
+        log(SetShortsTracker { shorts_tracker });
+        log(SetVault { vault });
+        log(SetRusdContract { rusd_contract });
+        log(SetCooldownDuration { cooldown_duration });
+    }
+
+    /*
+          ____     _       _           _       
+         / / /    / \   __| |_ __ ___ (_)_ __  
+        / / /    / _ \ / _` | '_ ` _ \| | '_ \ 
+       / / /    / ___ \ (_| | | | | | | | | | |
+      /_/_/    /_/   \_\__,_|_| |_| |_|_|_| |_|                         
+    */
+    #[storage(read, write)]
+    fn upgrade_and_withdraw(
+        asset: AssetId,
+        new_rlp_contract_receiver: ContractId,
+    ) {
```

Comment:
this feels a bit silly considering we can only withdraw one asset, we should give a way to withdraw all assets in this case right?

Reply:
rlp is removed